File: RecoverBinarySearchTree.cpp
// http://www.cnblogs.com/TenosDoIt/p/3445682.html
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void recoverTree(TreeNode *root) {
        //first,second 分别指向两个错误的节点，parent保存中序访问中当前节点的前驱
          TreeNode *first = NULL, *second = NULL, *parent = NULL;
          TreeNode *current = root, *pre = NULL;
          while(current != NULL)
          {                 
                if(current->left == NULL)
                {
                      if(parent != NULL)
                      {
                          if(first == NULL && current->val < parent->val)
                            first = parent;
                          else if(first && !second && current->val > first->val)
                            second = parent;
                      }
                      parent = current;
                      current = current->right;      
                }    
                else
                {
                      /* Find the inorder predecessor of current */
                      pre = current->left;
                      while(pre->right != NULL && pre->right != current)
                        pre = pre->right;
                        
                      if(pre->right == NULL)
                      {     /* Make current as right child of its inorder predecessor */
                            pre->right = current;
                            current = current->left;
                      }
                      else 
                      {
                            /* Revert the changes made in if part to restore the original 
                            tree i.e., fix the right child of predecssor */ 
                            //这里parent肯定不等于NULL
                            if(first == NULL && current->val < parent->val)
                                first = parent;
                            else if(first && !second && current->val > first->val)
                                second = parent;
                            parent = current;
                            
                            pre->right = NULL;
                            current = current->right;      
                      } 
                }
          } 
        if(first != NULL)
        {
            if(second == NULL)second = parent;//树{0,1}就可能出现这种情况
            int tmp = first->val;
            first->val = second->val;
            second->val = tmp;
        }
    }
};

File: jiuzhang/recover-binary-search-tree/recover-binary-search-tree.java
/**
 * 本代码由九章算法编辑提供。没有版权欢迎转发。
 * - 九章算法致力于帮助更多中国人找到好的工作，教师团队均来自硅谷和国内的一线大公司在职工程师。
 * - 现有的面试培训课程包括：九章算法班，系统设计班，九章强化班，Java入门与基础算法班，
 * - 更多详情请见官方网站：http://www.jiuzhang.com/
 */

/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    private TreeNode firstElement = null;
    private TreeNode secondElement = null;
    private TreeNode lastElement = new TreeNode(Integer.MIN_VALUE); 
    
    private void traverse(TreeNode root) {
        if (root == null) {
            return;
        }
        traverse(root.left);
        if (firstElement == null && root.val < lastElement.val) {
            firstElement = lastElement;
        }
        if (firstElement != null && root.val < lastElement.val) {
            secondElement = root;
        }
        lastElement = root;
        traverse(root.right);
    }
    
    public void recoverTree(TreeNode root) {
        // traverse and get two elements
        traverse(root);
        // swap
        int temp = firstElement.val;
        firstElement.val = secondElement.val;
        secondElement.val = temp;
    }
}

